mne/decoding/xdawn.py

```diff
@@ -0,0 +1,206 @@
+    mne.preprocessing.Xdawn
+    CSP
+
+    References
+    ----------
+    [1] Rivet, B., Souloumiac, A., Attina, V., & Gibert, G. (2009). xDAWN
+    algorithm to enhance evoked potentials: application to brain-computer
+    interface. Biomedical Engineering, IEEE Transactions on, 56(8), 2035-2043.
+
+    [2] Rivet, B., Cecotti, H., Souloumiac, A., Maby, E., & Mattout, J. (2011,
+    August). Theoretical analysis of xDAWN algorithm: application to an
+    efficient sensor selection in a P300 BCI. In Signal Processing Conference,
+    2011 19th European (pp. 1382-1386). IEEE.
+    """
+
+    def __init__(self, n_components=2, n_chan=None, signal_cov=None,
+                 reg=None):
+        """init xdawn."""
+        if n_chan is None:
+            raise ValueError("n_chan cannot be none. Please provide the "
+                             "number of channels")
+        self.n_chan = n_chan
+        super(XdawnTransformer, self).__init__(n_components, signal_cov, reg)
+
+    def fit(self, X, y):
+        """Fit Xdawn from epochs.
+
+        Parameters
+        ----------
+        X : ndarray, shape(n_channels, n_times * n_freq)
+            Data of epochs.
+        y : ndarray shape(n_samples,)
+            Target values.
+
+        Returns
+        -------
+        self : XdawnTransformer instance
+            The XdawnTransformer instance.
+        """
+        from sklearn.preprocessing import LabelEncoder
+
+        if X.ndim != 2 or not isinstance(X, np.ndarray):
+            raise ValueError("X should be 2 dimensional ndarray")
+        if not isinstance(y, np.ndarray):
+            raise ValueError("Labels must be numpy array")
+
+        epochs_data = X.reshape(X.shape[0], self.n_chan, X.shape[1] //
+                                self.n_chan)
+        # Extract signal covariance
+        self._get_signal_cov(epochs_data)
+
+        # estimates evoked covariance
+        evokeds = dict()
+        toeplitz = dict()
+        classes = LabelEncoder().fit(y).classes_
+        shape = epochs_data.shape
+        for eid in classes:
+            mean_data = np.mean(epochs_data[eid].reshape(-1, shape[1],
+                                shape[2]), axis=0)
+            evokeds[eid] = mean_data
+            toeplitz[eid] = 1.0
+
+        for eid in classes:
+            data = np.dot(evokeds[eid], toeplitz[eid])
+            self.evokeds_cov_[eid] = _regularized_covariance(data, self.reg)
+
+        # estimates spatial filters
+        for eid in classes:
+            self._fit_xdawn(eid)
+
+        # store some values
+        self.event_id = classes
+        self.exclude = list(range(self.n_components, self.n_chan))
+        return self
+
+    def transform(self, X):
+        """Apply Xdawn dim reduction.
+
+        Parameters
+        ----------
+        X : ndarray, shape(n_channels, n_times * n_freq)
+            data of epochs.
+
+        Returns
+        -------
+        X : ndarray, shape (n_epochs, n_components * event_types * n_times)
```

Comment:
In the dropbox doc I was suggesting of outputing `(n_epochs, n_components * event_types, n_times)` so that we stick to an Epochs dimensionality and thus easily pipe steps:
```
make_pipeline(
    XdawnTransformer(n_component=3, n_chan=64),  # output 3D
    SpatialFilter(3),  # => no need to initalize n_chan (which is good because it depends on the previous pipe step
    Vectorizer(),  # ensure 2D before sklearn step(s)
    LogisticRegression())  # apply sklearn step(s)